FirebaseTestProject/Models/Post.swift

```diff
@@ -13,13 +13,15 @@ struct Post: FirebaseConvertable {
     let author: String
     let text: String
     let id: UUID
+    var imageURL: String
     let timestamp: Date
     
-    init(title: String, text: String, author: String) {
+    init(title: String, text: String, author: String, imageURL: String = "no image") {
```

Comment:
Is it possible to have the `imageURL` be of type `String?` instead?   That way all posts wouldn't need to have images.  That'd be helpful because learners would first make posts without images, then add the images in and it'd be nice for their earlier posts to still work.

Reply:
Just to give my opinion, I would make the `imageURL` an optional value as well. It’s also a matter of teaching best practices: making use of optionals and other type safety features in Swift rather than stringly typing. 

LMK: What do we think about using a `URL?` instead of a `String?` for the `imageURL`? I would call this a best practice as well, but I understand it can be a little more cumbersome.

Reply:
The problem is saving it to firebase. During the encoding/decoding, firebase can't handle a nil object, specifically when it decodes to present the PostList.
We can have the student create the entire model so they can visualize a finished product and then build functionality as we go along.
To John's point, I started with a URL data type, but once again was limited by firebase data types.

Let me know y'alls thoughts.

Reply:
Ah, that makes sense! Thanks for letting us know about that.

I’d still be wary of letting limitations from Firebase leak into our codebase. What do we think about using DTOs (data transfer objects) to bridge that gap? It would help us write more Swifty code, but it comes at the expense of more boilerplate—and we’d probably want a lesson to explain it.

Reply:
@john-royal & @leftskidlow - I was talking to Ben about a potential solution and I think a quick solution to some of those problems would be the following:

It could be either `URL` or `String` -- preference here. More importantly, if you make a value in the model an Optional, if it can't encode that value in the initializer, it will just set the value to `nil`. This would solve the problem that as the application grows and older posts don't contain say an `imageURL`, you'd just make sure to have UI checks to cover the case that it doesn't exist. For `imageURL` you'd either show an image or maybe show a random default image (kind of like how Slack or Github assigns you a random image/color for your profile pic by default from a pool of images).

At this point, we won't have to be concerned with HOW the application grows, we just need to be concerned that if a Post doesn't contain a specific set of information, we simply won't include it and need to build the Post in a different manner.

I think this (hopefully) removes some of the boilerplate code and it is technically safe practice to have validation checks anyway, we are just adding a few in the UI incase certain properties exist or not.

Reply:
I think I agree with @Evilsmaher that we’ll want to store an optional value and handle both cases (value exists vs. value is `nil`) in our UI. However, Firebase can’t handle `nil` values.

To deal with Firebase, I agree with @leftskidlow’s solution: to include an `imageURL` property for each post, regardless of whether that post actually has an image attached. My preference would be to store an empty string instead of "no image", but I could be convinced otherwise.

Within our app, I have another idea: make the `imageURL` property a `URL?`, and use custom encoding/decoding logic to bridge the gap between the `String` in Firebase and the `URL?` in our app. I know this adds boilerplate, but keep in mind that the `Post` will likely need custom encoding/decoding logic regardless because of the `isFavorite` property. Here are some code snippets to show how that could work:

```swift
// Encoding
if let imageURL = imageURL {
    try container.encode(imageURL.absoluteString, forKey: .imageURL)
} else {
    try container.encode("", forKey: .imageURL)
}

// Decoding
let imageURLString = try values.decode(String.self, forKey: .imageURL)
imageURL = URL(string: imageURLString)
```

I would prefer this over the alternative (storing a `String` and working around it elsewhere in the application), but I could be convinced otherwise. I’d definitely like to hear everyone’s thoughts.